[SR. NO. 21] Fix Face Recognition Performance Bottleneck with Smart Clustering and Database Enhancements

priyankeshh opened the pull request:
## Overview
This PR addresses the critical performance bottleneck in face recognition clustering while introducing comprehensive database optimizations and monitoring capabilities.

## Core Changes

### Performance Optimizations
- Implemented smart batch processing for face embeddings
- Added incremental clustering strategy
- Optimized memory usage through efficient data structures
- Introduced database indexing for faster queries

### Database Enhancements
```python
# Key database improvements:
1. Context-managed connections
2. Batch operations support
3. Automated cleanup routines
4. Query optimization with GROUP_CONCAT
5. Performance monitoring utilities
```

### Memory Management
- Implemented configurable batch sizes
- Added memory-aware processing
- Optimized database operations for large datasets

## Technical Details

### Database Operations
```python
def batch_insert_face_embeddings(
    embeddings_batch: List[Tuple[str, List[np.ndarray]]],
    batch_size: int = 1000
) -> None:
    """
    Efficient batch processing with memory optimization
    """
```

### Performance Monitoring
```python
def get_database_stats() -> dict:
    """
    Real-time performance metrics and database health monitoring
    """
```

## 📊 Performance Test Results

![image](https://github.com/user-attachments/assets/d446837e-d8d3-4605-83d1-d3f862cd8ca4)


### Key Performance Metrics

![image](https://github.com/user-attachments/assets/d41cd163-5129-40f9-b796-338aaa70b50b)


### Test Configuration
- 100 images
- 5 embeddings per image
- 128-dimensional embeddings
- 1,000 batch size

## Breaking Changes
None. All changes are backward compatible.

## Checklist
- [x] Performance bottleneck resolved (98.1% improvement)
- [x] Database optimizations implemented
- [x] Memory management improved (minimal 3.5% increase)
- [x] Test results documented
- [x] Code reviewed

Fixes #350

priyankeshh committed:
- Introduced a ProgressTracker class to monitor long-running operations.
- Updated the detect_faces function to support progress tracking.
- Modified settings to define database paths and ensure necessary directories are created.
- Added new HTML and SVG files for frontend assets.

fixed merge conflicts

priyankeshh committed:
Add performance tests for embedding insertion and configure pytest for asyncio

priyankeshh committed:
Add memory usage tracking to embedding insertion performance test

priyankeshh committed:
Enhance performance test output with improved logging and detailed summary

priyankeshh committed:
Updated the PR with the codebase

priyankeshh committed:
style: format code with black and ruff

priyankeshh committed:
fix: update patch target for album description edit test

priyankeshh committed:
Resolved merge conflicts

priyankeshh committed:
Refactor image deletion logic and improve response formatting

- Removed unused variable `is_from_device` from `delete_image` function.
- Prefixed `is_from_device` in `delete_multiple_images` with an underscore for clarity.
- Simplified error handling for file not found cases.
- Enhanced response formatting in `delete_multiple_images` and `get_class_ids` functions.
- Added a blank line in `facetagging.py` for consistency.

priyankeshh committed:
fixed pytest

priyankeshh committed:
Added ruff and black

priyankeshh commented:
@Pranav0-0Aggarwal Would you please be able to review this PR when you have a moment? I've been keeping it up to date with the main branch, and it's ready for final review. Thank you !

priyankeshh committed:
Removed AI Generated content

Jibesh10101011 commented:
Hey @priyankeshh, why does this method batch_insert_face_embeddings ?

```py
def batch_insert_face_embeddings(
    embeddings_batch: List[Tuple[str, List[np.ndarray]]],
    batch_size: int = 1000
) -> None:
    """
    Efficient batch processing with memory optimization
    """
```
As In the actual application, the face-tagging algorithm processes only a single image_path when a new image is added:
[GitHub Link](https://github.com/AOSSIE-Org/PictoPy/blob/main/backend/app/routes/images.py)
![image](https://github.com/user-attachments/assets/11b526e7-b07c-4a64-b7fb-3d75683a534b)


So, at the moment, there doesn’t seem to be a use case for batch_insert_face_embeddings.

On what basis, then, were these 📊 Performance Test Results & Key Performance Metrics derived?

priyankeshh commented:
Hey @Jibesh10101011 

So about the batch insertion - you're right that the current flow processes images one by one. The thing is, I noticed while testing that even though we process images individually, the bottleneck happens when these operations stack up.

What was happening is that after every few images (around 5-6), the system would trigger a full reclustering operation with DBSCAN. This gets exponentially slower as more images are added to the system. So even though we're adding images one at a time, the clustering was becoming a major bottleneck.

The batch insertion function is partly forward-looking - I'm planning to update the scheduler in a follow-up PR to better handle bulk operations. But even with the current flow, it helps optimize how we handle the database connections and reduces the clustering overhead.

For the performance metrics, I wrote a test script that simulates adding multiple images to the system. It basically replicates what happens when users import a folder or when multiple images get processed in sequence. The 98% improvement comes mainly from avoiding unnecessary reclustering operations and optimizing the database interactions.
You can see the [test_performance.py](https://github.com/AOSSIE-Org/PictoPy/pull/354/files#diff-1cd89ca4a52365d2e3e85deb912da7a325030cb74d25e98b8127536a4f5af131)

Jibesh10101011 commented:
> The batch insertion function is partly forward-looking - I'm planning to update the scheduler in a follow-up PR to better handle bulk operations. But even with the current flow, it helps optimize how we handle the database connections and reduces the clustering overhead.
> 

I see your point about optimizing database connections, but this batch insertion approach is only beneficial when processing multiple images at once. However, in our current architecture, each image is first checked for its type—if it's a 'person,' only then does it proceed to clustering. This check happens specifically when handling @router.route("/images/add-folder").

Now, if I modify the architecture as per your suggestion, let’s analyze the impact. Suppose we have 1000 images, and 500 of them are of type 'person.' Your approach suggests inserting these 500 images in a batch. However, after insertion, all these images still need to be compared with existing face embeddings for clustering. This means the linear expansion issue persists—batching doesn't eliminate the fundamental scaling problem.

priyankeshh commented:
Oh wait, I see what you're saying now. You're absolutely right - I got so focused on optimizing the database operations that I missed the bigger bottleneck in the clustering algorithm itself. 

I fixed the problem locally and I will updating this PR soon

priyankeshh committed:
Implement batch processing for face detection and clustering; add tests for incremental clustering performance

priyankeshh committed:
Added ruff check and black check

priyankeshh committed:
Remove unused Path import and clarify project root path addition in tests

priyankeshh committed:
fixed linting

priyankeshh commented:
I've now implemented incremental clustering to address the performance bottleneck:

**Before:**

1. Every new face embedding triggered a full DBSCAN reclustering
2. O(n²) complexity made processing slow with larger datasets
3. No batch processing support for multiple faces

**Now:**

1. New approach avoids rerunning DBSCAN on the entire dataset for each new image
2. Uses vectorized distance calculations to efficiently compare new embeddings only to existing clusters
3. Only assigns new cluster IDs when necessary
4. Added proper batch processing support in add_faces_batch()

Added test file:

1. Created [tests/test_incremental_clustering.py](https://github.com/AOSSIE-Org/PictoPy/pull/354/commits/68c8ec66bad921620f954503c33834c33c865012#diff-a1e2128d75d6afa705c6b65124b7d10e9b68888be571b66c85cd8e5c6851f72b) to benchmark performance

- Tests Result:
![image](https://github.com/user-attachments/assets/bd1b9278-1145-4347-b13f-c63d7642deff)



The changes maintain clustering accuracy while dramatically improving performance. This addresses the O(n²) complexity problem you pointed out, making the application responsive even with large image collections.

priyankeshh commented:
@Pranav0-0Aggarwal  can you take a look at this PR?
